bgpd/bgp_route.c

```diff
@@ -2221,39 +2239,64 @@ void bgp_process_queue_init(void)
 		}
 	}
 
-	bm->process_main_queue->spec.workfunc = &bgp_process_main;
+	bm->process_main_queue->spec.workfunc = &bgp_process_wq;
 	bm->process_main_queue->spec.del_item_data = &bgp_processq_del;
 	bm->process_main_queue->spec.max_retries = 0;
 	bm->process_main_queue->spec.hold = 50;
 	/* Use a higher yield value of 50ms for main queue processing */
 	bm->process_main_queue->spec.yield = 50 * 1000L;
 }
 
+static struct bgp_process_queue *bgp_process_queue_work(struct work_queue *wq,
+							struct bgp *bgp)
+{
+	struct bgp_process_queue *pqnode;
+
+	pqnode = XCALLOC(MTYPE_BGP_PROCESS_QUEUE, sizeof(struct bgp_process_queue));
+
+	/* unlocked in bgp_processq_del */
+	pqnode->bgp = bgp_lock(bgp);
+	STAILQ_INIT(&pqnode->pqueue);
+
+	work_queue_add(wq, pqnode);
+
+	return pqnode;
+}
+
 void bgp_process(struct bgp *bgp, struct bgp_node *rn, afi_t afi, safi_t safi)
 {
+#define ARBITRARY_PROCESS_QLEN		10000
+	struct work_queue *wq = bm->process_main_queue;
 	struct bgp_process_queue *pqnode;
 
 	/* already scheduled for processing? */
 	if (CHECK_FLAG(rn->flags, BGP_NODE_PROCESS_SCHEDULED))
 		return;
 
-	if (bm->process_main_queue == NULL)
+	if (wq == NULL)
 		return;
 
-	pqnode = XCALLOC(MTYPE_BGP_PROCESS_QUEUE,
-			 sizeof(struct bgp_process_queue));
-	if (!pqnode)
-		return;
+	/* Add route nodes to an existing work queue item until reaching the
+	   limit only if is from the same BGP view and it's not an EOIU marker */
+	if (work_queue_item_count(wq)) {
+		struct work_queue_item *item = work_queue_last_item(wq);
+		pqnode = item->data;
 
-	/* all unlocked in bgp_processq_del */
+		if (CHECK_FLAG(pqnode->flags, BGP_PROCESS_QUEUE_EOIU_MARKER) ||
+		    pqnode->bgp != bgp || pqnode->queued >= ARBITRARY_PROCESS_QLEN)
+			pqnode = bgp_process_queue_work(wq, bgp);
+	} else
+		pqnode = bgp_process_queue_work(wq, bgp);
+
+	/* all unlocked in bgp_process_wq */
 	bgp_table_lock(bgp_node_table(rn));
-	pqnode->rn = bgp_lock_node(rn);
-	pqnode->bgp = bgp;
-	bgp_lock(bgp);
-	pqnode->afi = afi;
-	pqnode->safi = safi;
-	work_queue_add(bm->process_main_queue, pqnode);
+	rn->afi = afi;
+	rn->safi = safi;
 	SET_FLAG(rn->flags, BGP_NODE_PROCESS_SCHEDULED);
+
+	STAILQ_INSERT_TAIL(&pqnode->pqueue, bgp_lock_node(rn), pq);
```

Comment:
a lock inside a macro is generally considered a bad idea.  Please move the lock to before call.